### PR TITLE
[Finetune] replace fine-tuning DefaultTrainer with transformers.Trainer

### PR DESCRIPTION
**Pull request by harborn**
refactoring: replace fine-tuning DefaultTrainer with transformers.Trainer

this update will:
1. disable DefaultTrainer, which only contains only a very small subset of training functionality.
2. enable completely arguments support for different training task.

### CODE REVIEW

File: llm_on_ray/finetune/finetune.py [hunk 1]
```diff
@@ -193,92 +242,43 @@ def train_func(config: Dict[str, Any]):
...
-            "checkpoint": {
-                "root_path": config["General"].get("checkpoint_dir", None),
-            },
-        }
-    )
-
-    try:
-        common.logger.info("trainer prepare start")
-        model.training = True
-        trainer.prepare(model, tokenizer, datasets, optimizer, accelerator)
```
Comment (minmingzhu) [inline llm_on_ray/finetune/finetune.py:269]: Does remove prepare function in default_trainer.py?

File: llm_on_ray/finetune/finetune.py [hunk 2]
```diff
@@ -193,92 +242,43 @@ def train_func(config: Dict[str, Any]):
...
+        tokenizer=tokenizer, mlm=False, return_tensors="pt", pad_to_multiple_of=8
     )
 
-    accelerate_mode = config["Training"]["accelerate_mode"]
-    if accelerate_mode == "FSDP":
-        fsdp_plugin = FullyShardedDataParallelPlugin(
-            state_dict_config=FullStateDictConfig(offload_to_cpu=False, rank0_only=False),
-            optim_state_dict_config=FullOptimStateDictConfig(
-                offload_to_cpu=False, rank0_only=False
-            ),
```
Comment (minmingzhu) [inline llm_on_ray/finetune/finetune.py:210]: transformers.Trainer how to distinguish FSDP and Deepspeed?

File: llm_on_ray/finetune/finetune.py [hunk 3]
```diff
@@ -193,92 +242,43 @@ def train_func(config: Dict[str, Any]):
         }
     )
 
-    optimizer = common.optimizer.Optimizer.registory.get("DefaultOptimizer")()(
-        model,
-        config={
-            "name": config["Training"]["optimizer"],
-            "config": {"lr": config["Training"]["learning_rate"]},
-        },
```
Comment (KepingYan) [inline llm_on_ray/finetune/finetune.py:201]: Why remove optimizer?

File: llm_on_ray/finetune/finetune.py [hunk 4]
```diff
@@ -193,92 +242,43 @@ def train_func(config: Dict[str, Any]):
...
+        from transformers import Trainer, TrainingArguments
+
+        training_args = convert_to_training_args(TrainingArguments, config)
+        trainer = Trainer(
+            model=model,
+            args=training_args,
+            train_dataset=tokenized_datasets["train"],
+            eval_dataset=tokenized_datasets["validation"],
+            tokenizer=tokenizer,
+            data_collator=data_collator,
```
Comment (KepingYan) [inline llm_on_ray/finetune/finetune.py:259]: I think we should pass optimizer parameter into the Trainer according to the user's configuration.

File: llm_on_ray/finetune/finetune.py [hunk 5]
```diff
@@ -130,6 +121,52 @@ def get_accelerate_environment_variable(config: Dict[str, Any]) -> dict:
...
+        "save_strategy": save_strategy,
+        "bf16": config["Training"]["mixed_precision"] == "bf16",
+        "num_train_epochs": config["Training"]["epochs"],
+        "per_device_train_batch_size": config["Training"]["batch_size"],
+        "per_device_eval_batch_size": config["Training"]["batch_size"],
+        "learning_rate": config["Training"]["learning_rate"],
+        "logging_steps": config["Training"]["logging_steps"],
+        "lr_scheduler_type": config["Training"]["lr_scheduler"],
+        "weight_decay": config["Training"]["weight_decay"],
+        "gradient_accumulation_steps": config["Training"]["gradient_accumulation_steps"],
```
Comment (carsonwang) [inline llm_on_ray/finetune/finetune.py:142]: Do we have default values for all these configurations in `config`? Previously we wrote `config["Training"].get("gradient_accumulation_steps", 1)`

File: llm_on_ray/finetune/finetune.py [hunk 6]
```diff
@@ -339,27 +339,22 @@ def main(external_config=None):
                 "FI_PROVIDER": "tcp",
             }
         }
-        accelerate_env_vars = get_accelerate_environment_variable(config)
-        runtime_env["env_vars"].update(accelerate_env_vars)
+        # accelerate_env_vars = get_accelerate_environment_variable(config)
+        # runtime_env["env_vars"].update(accelerate_env_vars)
```
[the remark this answers is not in the document]
Comment (harborn) [reply llm_on_ray/finetune/finetune.py:343]: Yes, no longer needed. removed.

File: llm_on_ray/finetune/finetune.py [hunk 7]
```diff
@@ -193,92 +242,43 @@ def train_func(config: Dict[str, Any]):
...
+
+        training_args = convert_to_training_args(TrainingArguments, config)
+        trainer = Trainer(
+            model=model,
+            args=training_args,
+            train_dataset=tokenized_datasets["train"],
+            eval_dataset=tokenized_datasets["validation"],
+            tokenizer=tokenizer,
+            data_collator=data_collator,
         )
```
[the remark this answers is not in the document]
Comment (harborn) [reply llm_on_ray/finetune/finetune.py:260]: checkpoint and model result saving are control by argument in `training_args`, there are options:
`save_only_model`, `save_strategy`, `save_steps`, `output_dir`, `save_steps`, etc.
all this options can be used to match different need of all kinds saving and loading models and checkpoints.
For more details, please see here.[https://huggingface.co/docs/transformers/v4.40.2/en/main_classes/trainer#transformers.TrainingArguments](url)

File: llm_on_ray/finetune/finetune.py (hunk 3 above)
Comment (harborn) [reply llm_on_ray/finetune/finetune.py:201]: optimizer will create in `transformers.Trainer` or `optimum.habana.transformers.GaudiTrainer`

File: llm_on_ray/finetune/finetune.py (hunk 4 above)
Comment (harborn) [reply llm_on_ray/finetune/finetune.py:259]: optimizer's parameters are parts of `TrainingArguments`, and optimizer will be create before calling epoch-step-loop in Trainer.train()

File: llm_on_ray/finetune/finetune.py [hunk 8]
```diff
@@ -130,6 +121,52 @@ def get_accelerate_environment_variable(config: Dict[str, Any]) -> dict:
     return mode_env_vars[device][accelerate_mode]
 
 
+def convert_to_training_args(cls, config):
+    device = config["Training"]["device"]
+    accelerate_mode = config["Training"]["accelerate_mode"]
+    checkpoint_dir = config["General"]["checkpoint_dir"]
```
[the remark this answers is not in the document]
Comment (harborn) [reply llm_on_ray/finetune/finetune.py:127]: removed.

File: llm_on_ray/finetune/finetune.py (hunk 5 above)
Comment (harborn) [reply llm_on_ray/finetune/finetune.py:142]: yes, this option has default value 1 in finetune_config.py

File: llm_on_ray/finetune/finetune.py [hunk 9]
```diff
@@ -130,6 +121,52 @@ def get_accelerate_environment_variable(config: Dict[str, Any]) -> dict:
...
+    if accelerate_mode == "DEEPSPEED":
+        args.update({"deepspeed": config["Training"]["deepspeed_config_file"]})
+
+    # set attr for Intel Gaudi
+    if device == "hpu":
+        args.update({"use_habana": True})
+        args.update({"use_lazy_mode": config["Training"]["hpu_execution_mode"] == "lazy"})
+        args.update({"pipelining_fwd_bwd": True})
+        args.update({"throughput_warmup_steps": 3})
+        args.update({"adam_epsilon": 1e-8})
```
[the remark this answers is not in the document]
Comment (harborn) [reply llm_on_ray/finetune/finetune.py:165]: I removed this hard coded value. I am thinking should add those options to our yaml config file.

File: llm_on_ray/common/dataprocesser/general_processer.py [hunk 10]
```diff
@@ -176,13 +173,22 @@ def group_texts(examples):
                 desc=f"Grouping texts in chunks of {block_size}",
             )
 
+        return tokenized_datasets
+
+    def convert_dataset(self, tokenizer, dataset):
```
[the remark this answers is not in the document]
Comment (harborn) [reply llm_on_ray/common/dataprocesser/general_processer.py:178]: updated!

File: llm_on_ray/finetune/finetune.py (hunk 2 above)
Comment (harborn) [reply llm_on_ray/finetune/finetune.py:210]: for FSDP training, we should change TrainingArguments's option as following:
`fsdp_config` with a json config file: fsdp_config.json
`fsdp auto_wrap` set True

File: llm_on_ray/finetune/finetune.py (hunk 4 above)
Comment (KepingYan) [reply llm_on_ray/finetune/finetune.py:259]: But I don’t see the optimizer configured in TrainingArguments. How does user set optimizer parameter?

File: llm_on_ray/finetune/finetune.py [hunk 11]
```diff
@@ -322,44 +323,27 @@ def main(external_config=None):
...
-        ccl_worker_count = num_training_workers
+        pass
 
     if not ray.is_initialized():
         runtime_env = {
             "env_vars": {
-                "OMP_NUM_THREADS": str(resources_per_worker["CPU"]),
-                "CCL_ZE_IPC_EXCHANGE": "sockets",
-                "CCL_WORKER_COUNT": str(ccl_worker_count),
-                "CCL_LOG_LEVEL": "info",
```
Comment (KepingYan) [inline llm_on_ray/finetune/finetune.py:336]: Why are these ccl configurations no longer needed?

**Review by KepingYan — APPROVED** [review]

**Comment by KepingYan** [comment]
LGTM

**Review by carsonwang — APPROVED** [review]

**Pull request merged**